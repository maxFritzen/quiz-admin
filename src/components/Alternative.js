import React from 'react';
import { connect } from 'react-redux';
import { alternativeInput, validateInput } from '../actions';

class Alternative extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      error: false,
      errorMessage: 'error'
    };
  }
  onFocus = () => {
    this.setState({
      error: false
    });
  }

  onBlur = () => {
    if(!this.props.value) {
      this.setState({
        error: true,
        errorMessage: 'Please enter value'
      });
    }
  }

  onChange = (e, index) => {
      if(this.props.error.questions[this.props.questionIndex]) {
          // This one validates after input, making sure the warning disappears
          // when OK.
          this.props.validateInput('alternative', e.target.value, index, this.props.questionIndex);
      } else {
      // Validation happens in local state, and only when onBlur.
      this.props.alternativeInput(e.target.value, index, this.props.questionIndex);
    }
  }

  onFocus = () => {
    this.setState({
      error: false
    });
  }

  render() {
    return (
      <div>
        <input
          value={this.props.value}
          onChange={(e) => this.onChange(e, this.props.index)}
          placeholder="Alternative"
          onFocus={this.onFocus}
          onBlur={this.onBlur}
        />
        {this.state.error && this.state.errorMessage}

      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  error: state.validate.error,
  form: state.form
});

const mapDispatchToProps = (dispatch) => ({
  alternativeInput: (value, index, questionIndex) => dispatch(alternativeInput(value, index, questionIndex)),
  validateInput: (dispatchToValidate, value, index, questionIndex) => dispatch(validateInput(dispatchToValidate, value, index, questionIndex)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Alternative);

// const Alternative = (props) => (
//   <div>
//     <input
//       value={props.value}
//       onChange={(e) => props.onChange(e, props.index)}
//       placeholder="Alternative"
//       // onFocus={() => props.onFocus('alt')}
//       // onBlur={() => props.onBlur('alt')}
//     />
//     {/* {props.blur && props.error} */}
//   </div>
//
// );
//
// export default Alternative;
