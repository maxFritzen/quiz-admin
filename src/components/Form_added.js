import React from 'react';

const Form_added = () => (
  <div>
    Added form!
  </div>
);

export default Form_added;
