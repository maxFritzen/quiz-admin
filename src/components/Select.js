import React from 'react';

const Select = (props) => (
  <select value={props.selectValue} onChange={props.handleChange}>
    <option>Choose Correct Answer</option>
    {props.alternativeValues.map((value, index) => {
      return <option key={index} value={value}>{value}</option>
    })}
  </select>
);

export default Select;
