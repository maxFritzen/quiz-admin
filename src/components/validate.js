const validate = (values) => {
  const errors = {};
  if(!values.title) {
    errors.title = 'Please enter a title'
  }
  return errors;
}

export default validate;
