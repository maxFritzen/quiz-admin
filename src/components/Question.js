import React from 'react';
import Alternative from './Alternative';
import Select from './Select';




export default class Question extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectValue: '',
      alternatives: ['','']
    }

  }

  handleInputChange = (e, index) => {
    const newArray = this.state.alternatives;
    newArray[index] = e.target.value;
    this.setState({
     alternatives: newArray
   }, () =>   console.log(this.state.alternatives));
  }

  handleSelectChange = (e) => {
    this.setState({
      selectValue: e.target.value
    });
  }

  handleAddAlternative = () => {
    const name = this.state.x;
    const newArray = this.state.alternatives;
    newArray.push('');
    console.log(newArray);
    this.setState({
      alternatives: newArray,
      x: this.state.x +1
    });
  };

  remove = (index) => {
      // Filter out index to be removed
    this.setState((prevState) => ({
           alternatives: prevState.alternatives.filter((i) => i !== this.state.alternatives[index])
         }));

  }

  render() {
  const alternatives = this.state.alternatives;

    return (
      <div>
        <h4>Question GÃ–R INPUT</h4>
        <ul>
          {alternatives.map((alternative, index) => (
              <li key={index}>
                 <Alternative
                    index={index}
                    value={this.state.alternatives[index]}
                    onChange={this.handleInputChange}
                 />
                <button type="button" onClick={() => this.remove(index)}>remove alternative</button>
              </li>
          ))}
        </ul>
        <button type="button" onClick={this.handleAddAlternative}>Add Alternative</button>
        <Select
          selectValue={this.state.selectValue}
          handleChange={this.handleSelectChange}
          alternatives={this.state.alternatives}
        />
      </div>
    );
  }
}
