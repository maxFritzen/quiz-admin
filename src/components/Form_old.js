import React from 'react';
import { Field, FieldArray, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import validate from './validate';
import { createQuiz } from '../actions';

const renderField = ({ input, label, type, meta: { touched, error } }) => (
  <div>
    <label>{label}</label>
    <div>
      <input
        {...input}
        type={type}
        placeholder={label}
      />
    </div>
  </div>
);

const renderQuestions = ({ fields, meta: { error, submitFailed } }) => (
  <ul>
    <li>
      <button type="button" onClick={() => fields.push({})}>
        Add Question
      </button>
      {submitFailed && error && <span>{error}</span>}
    </li>
    {fields.map((question, index) => (

      <li key={index}>
        <button
          type="button"
          title="Remove Question"
          onClick={() => fields.remove(index)}
        >
        Remove Question
        </button>
        <h3>Question #{index + 1} </h3>
        <Field
          name={`${question}.question`}
          type="text"
          component={renderField}
          label="Question"
        />
        <FieldArray name={`${question}.alternatives`} component={renderAlternatives} />


        <Field name={`${question}.correct`} component="select">
          <option></option>


            <option value="00ff00">{fields.question.question}</option>
            <option value="0000ff">Blue</option>
        </Field>
      </li>
    ))}
  </ul>
);

export const renderAlternatives = ({ fields, meta: { error } }) => (
  <ul>
    <li>
      <button type="button" onClick={() => fields.push()}>
        Add Alternative
      </button>
    </li>
    {fields.map((alternative, index) => (
      <li key={index}>
        <button
          type="button"
          title="Remove Alternative"
          onClick={() => fields.remove(index)}
          >
          Remove Alternative
        </button>
        <Field
          name={alternative}
          type="text"
          component={renderField}
          label={`Alternative #${index + 1}`}
        />
      </li>
    ))}
  </ul>
);

const onSubmit = (values) => {
  console.log('values:', values);
  // createQuiz(values);
}


export const Form = (props) => {
  const { handleSubmit } = props;
  return (
    <form onSubmit={handleSubmit(onSubmit.bind(this))} className="form-horizontal">
      <Field
        name="quizName"
        type="text"
        component={renderField}
        label="Quiz Title"
      />
      <FieldArray name="questions" component={renderQuestions} />
      <button type="submit" className="btn btn-primary">
        Submit
      </button>
    </form>
  );
}

export default reduxForm({
  form: 'quizForm',
  validate
})(
  connect(null, {createQuiz})(Form)
);
