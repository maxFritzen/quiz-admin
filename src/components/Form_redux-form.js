import React from 'react';
import { Field, FieldArray, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import validate from './validate';
import { createQuiz } from '../actions';

// class renderQuestions extends React.Component {
//   render() {
//     return (
//       <ul>
//          <li>
//            <button type="button" onClick={() => fields.push({})}>
//              Add Question
//            </button>
//            {submitFailed && error && <span>{error}</span>}
//          </li>
//          {fields.map((question, index) => (
//           <li key={index}>
//             <button
//               type="button"
//               title="Remove Question"
//               onClick={() => fields.remove(index)}
//             >
//             Remove Question
//             </button>
//             <h3>Question #{index + 1} </h3>
//             <Field
//               name={`${question}.question`}
//               type="text"
//               component={this.renderField}
//               label="Question"
//             />
//             <FieldArray name={`${question}.alternatives`} component={renderAlternatives} />
//
//
//             <Field name={`${question}.correct`} component="select">
//               <option></option>
//
//                 <option value="00ff00">asd</option>
//                 <option value="0000ff">Blue</option>
//             </Field>
//           </li>
//         ))}
//       </ul>
//     );
//   }
// }

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
    alternatives: []
    }
  }

  renderField = ({ input, label, type, meta: { touched, error } }) => (

    <div>
      <label>{label}</label>
      <div>
        <input
          {...input}
          type={type}
          placeholder={label}
        />
      </div>
    </div>
  );

  renderQuestions = ({ fields, meta: { error, submitFailed } }) => {
    const alternatives = 'test';
    return (
      <ul>
        <li>
          <button type="button" onClick={() => fields.push({})}>
            Add Question
          </button>
          {submitFailed && error && <span>{error}</span>}
        </li>
        {fields.map((question, index) => (

          <li key={index}>
            <button
              type="button"
              title="Remove Question"
              onClick={() => fields.remove(index)}
            >
            Remove Question
            </button>
            <h3>Question #{index + 1} </h3>
            <Field
              name={`${question}.question`}
              type="text"
              component={this.renderField}
              label="Question"
            />
            <FieldArray name={`${question}.alternatives`} component={this.renderAlternatives} />


            {/* <Field name={`${question}.correct`} component="select">
              <option></option>

                <option value="00ff00">{question[0]}</option>
                <option value="0000ff">Blue</option>
            </Field> */}
          </li>
        ))}
      </ul>
    )
  };

  renderAlternatives = ({ fields, meta: { error } }) => {

    return (
      <ul>
        <li>
          <button type="button" onClick={() => fields.push()}>
            Add Alternative
          </button>
        </li>
        {fields.map((alternative, index) => (
          <li key={index}>
            <button
              type="button"
              title="Remove Alternative"
              onClick={() => fields.remove(index)}
              >
              Remove Alternative
            </button>
            <Field
              name={alternative}
              type="text"
              component={this.renderField}
              label={`Alternative #${index + 1}`}
            />
            <Field
              name={`${alternative}`}
              type="checkbox"
              component="input"
              label="js"

            />
          </li>
        ))}

      </ul>
    );


  }

  onSubmit = (values) => {
   console.log('values:', values);
   // createQuiz(values);
 }


  render() {
    const { handleSubmit } = this.props;
    return (
      <form onSubmit={handleSubmit(this.onSubmit.bind(this))} className="">
        <h1>testing:{this.props.test}</h1>
        <Field
          name="quizName"
          type="text"
          component={this.renderField}
          label="Quiz Title"
        />
        <FieldArray name="questions" component={this.renderQuestions} />
        <button type="submit" className="btn btn-primary">
          Submit
        </button>
      </form>
    );
  }
}

const mapStateToProps = (state) => {
  test: state.quizForm
}

export default reduxForm({
  form: 'quizForm',
  validate
})(
  connect(mapStateToProps, {createQuiz})(Form)
);
