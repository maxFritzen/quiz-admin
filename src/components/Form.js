import React from 'react';
import { connect } from 'react-redux';
import Question from './Question';

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      numberOfQuestions: 2
    }
  }

  handleAddQuestion = () => (
    this.setState({
      numberOfQuestions: this.state.numberOfQuestions + 1
    })
  );

  onSubmit = (e) => {
    e.preventDefault();
    console.log('values:', e)
  }

  render() {
    const Questions = [];

    for (let i= 0; i < this.state.numberOfQuestions; i++) {
      Questions.push(<Question />);
    }
    return (
      <form onSubmit={this.onSubmit}>
        FORM
        Number of questions: {this.state.numberOfQuestions}
        {Questions.map((question, index) => {
          return <li key={index}>{question}</li>
        })}
        {/* {this.renderQuestions()} */}
        <button onClick={this.handleAddQuestion}>Add Question</button>
        <button type="submit">Save</button>
      </form>
    );
  }
}

export default connect()(Form);
