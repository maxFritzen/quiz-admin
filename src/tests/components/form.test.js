import React from 'react';
import { shallow } from 'enzyme';
import  { Form, renderAlternatives as RenderAlternatives}  from '../../components/Form';

const handleSubmit = jest.fn();
test('Should render Form component', () => {
  const wrapper = shallow(<Form handleSubmit={handleSubmit}/>);
  expect(wrapper).toMatchSnapshot();
});

test('Submit button calls handleSubmit', () => {
  const wrapper = shallow(<Form handleSubmit={handleSubmit}/>);
  wrapper.find('form').simulate('submit');
  expect(handleSubmit).toHaveBeenCalled();
});

const fields = [];
const meta = {error:'error'};
test('Should render renderAlternatives component', () => {
  const wrapper = shallow(<RenderAlternatives fields={fields} meta={meta}/>);
  expect(wrapper).toMatchSnapshot();
});
