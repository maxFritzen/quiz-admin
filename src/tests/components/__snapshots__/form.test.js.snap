// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Form component 1`] = `
<form
  className="form-horizontal"
>
  <Field
    component={[Function]}
    label="Quiz Title"
    name="quizName"
    type="text"
  />
  <FieldArray
    component={[Function]}
    name="questions"
  />
  <button
    className="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`Should render renderAlternatives component 1`] = `
<ul>
  <li>
    <button
      onClick={[Function]}
      type="button"
    >
      Add Alternative
    </button>
  </li>
</ul>
`;
